docs/tutorial/03-let-the-snake-move.md

```diff
@@ -78,8 +78,119 @@ end
 We'll also turn our `draw_object/2` function into an `draw_objects/2` function. Instead of just the snake object, we'll pass the whole state as the second argument.
 
 ```elixir
-defp draw_objects(graph, %{snake: %{body: snake}}) do
-  Enum.reduce(snake, graph, fn {x, y}, graph ->
+defp draw_objects(graph, %{snake: %{body: body}}) do
+  Enum.reduce(body, graph, fn {x, y}, graph ->
+    draw_tile(graph, x, y, fill: :dark_slate_gray)
+  end)
+end
+```
+
+
+- add direction to snake
+
+```elixir
+state = %{
+  width: num_tiles_width,
+  height: num_tiles_height,
+  snake: %{body: [{9, 9}, {10, 9}, {11, 9}], size: 5, direction: {1, 0}}
+}
+```
+
+- add function to move snake and return new state:
+
+```elixir
+defp move_snake(%{snake: snake} = state) do
+  %{body: body, size: size, direction: direction} = snake
+
+  # new head
+  [head | _] = body
+  new_head = move(state, head, direction)
+
+  # truncate body
+  new_body = Enum.take([new_head | body], size)
+
```

Comment:
Going through these lines of code again, I think this is the wrong approach. Let's say we start with
```body = [{9, 9}, {10, 9}, {11, 9}]``` . On the first iteration, we'll have `new_body = [{10, 9}, {9,9}, {10, 9}]`, which is not correct. The following iterations will be correct, though: `new_body =  [{11, 9}, {10,9}, {9, 9}]`. I think the correct approach would be to append a new tuple to the end of the list and use `Enum.take` to cut from the end of the list.
So that on the first iteration, `new_body = [{10, 9}, {11, 9}, {12, 9}]`. What do you think, @nickgnd @klappradla ?
Also we're using `size = 5`, but it should be 3, right?

Reply:
Oh, very very good points @LuisaAPF 🙈 

Ad **size**
Let's fully remove this from the code for now. We're not making proper use of it. And if we would use it, we should rather compute it from the length of the body map instead of hardcoding it in my opinion.

Ad **movement**
You're also perfectly right here 💯 
I just quickly tried it out and the results don't really make much sense (especially with us using the wrong size... 🙈).

It looks like we're mixing two possible approaches:
1. Only store the `head` and the `size` of the snake and draw all other dots based on the `size` and `direction`
1. Store the `body` and translate **all** dots when moving

I personally find approach no. 2 more natural for the game. We could simply recurse through the list and apply the same transformation to each dot. I think it would simplify the code a lot.

What's your opinion on that?


Oh, very good points!

First off: the **size**: Let's fully remove this from the code for now. We're not making any use of it until we count points if I remember the original tutorial correctly. We should wait until this step and then compute the size rather than hardcoding it.

For the body: you're absolutely right 👍. I

Reply:
@LuisaAPF Thanks for the good catch 👁 👏 

> Let's fully remove the size

👍 

> Only store the head and the size of the snake and draw all other dots based on the size and direction

 👎  I don't think this approach would work since it won't allow the snake to have multiple segments, but only one.

> Store the body and translate all dots when moving

👍 Yup, but we need to apply the transformation only to the head dot

Reply:
I agree with using the list size, instead of adding a `size` key. 

>  Yup, but we need to apply the transformation only to the head dot

I'm not sure I understand what you meant @nickgnd. I guess we should apply the transformation to all elements in the list, no?

Reply:
Forget it. I get it now :+1:

Reply:
Hm, I still don't fully get it 🙈. I thought recursively transforming **all** dots would be the way to go, ignoring anything size related.

But if you already have an idea, please just add it and push it. It will probably be easier to reason about it then 😉

Reply:
If I take the existing code and just adapt the `move_snake` function to this:
```elixir
defp move_snake(%{snake: snake} = state) do
  %{body: body, direction: direction} = snake
  new_body = Enum.map(body, &move(state, &1, direction))

  state
  |> put_in([:snake, :body], new_body)
end
```

Everything works perfectly fine and we get the right values.

One could of course adapt the functions for a nicer pipeline, etc. but the approach is essentially what I meant with recursively transforming each dot.

What do you think?

Reply:
If you want to try it out in the real world: I also quickly updated my _"snake notepad"_ to use this approach -> https://github.com/klappradla/snake/commit/4aec0eb8f3309a8af3fe8ba9ab6850da8ff90cb6

Reply:
I also thought about that. But that only works if you think in one dimension. As soon as you start thinking about the vertical axis, you see that it doesn't work. It will move the whole snake vertically, because we're applying the same transformation to all the tiles.

Reply:
Oh 🙈 - now I get the problem 💡 

Thanx a ton @LuisaAPF 🙏